File: src/nmath/quadratics.zig
const constants = @import("constants.zig");
const mathf = @import("mathf.zig");
const std = @import("std");

pub const NumSolutions = enum {
    Zero,
    One,
    Two,
};

pub const Solution = union(NumSolutions) {
    Zero: void,
    One: f32,
    Two: struct { f32, f32 },
};

pub fn solve(a: f32, b: f32, c: f32) Solution {
    const discriminant = b * b - 4 * a * c;
    return if (discriminant < -constants.epsilon)
        Solution{ .Zero = {} }
    else if (mathf.approxEq(discriminant, 0.0))
        Solution{ .One = (-b) / (2.0 * a) }
    else
        Solution{ .Two = .{
            ((-b) + @sqrt(discriminant)) / (2.0 * a),
            ((-b) - @sqrt(discriminant)) / (2.0 * a),
        } };
}

pub fn approxEq(a: Solution, b: Solution) bool {
    return switch (a) {
        .Zero => b == .Zero,

        .One => |x| switch (b) {
            .One => |y| mathf.approxEq(x, y),
            else => false,
        },

        .Two => |x| switch (b) {
            .Two => |y| (mathf.approxEq(x[0], y[0]) and mathf.approxEq(x[1], y[1])) or
                (mathf.approxEq(x[0], y[1]) and mathf.approxEq(x[1], y[0])),
            else => false,
        },
    };
}

test "quadratics:solve works" {
    const s = solve(1, -4, -8);
    const expected = Solution{ .Two = .{
        2.0 + 2.0 * @sqrt(3.0),
        2.0 - 2.0 * @sqrt(3.0),
    } };

    try std.testing.expect(approxEq(s, expected));
}

File: pkg/nmath/quadratics.go
package nmath

import "math"

type Solution []float64

func Solve(a, b, c float64) Solution {
	discriminant := b*b - 4*a*c
	if discriminant < -f64Epsilon {
		return Solution{}
	} else if ApproxEq(discriminant, 0.0) {
		return Solution{(-b) / (2.0 * a)}
	} else {
		return Solution{
			((-b) + math.Sqrt(discriminant)) / (2.0 * a),
			((-b) - math.Sqrt(discriminant)) / (2.0 * a),
		}
	}
}

func (a Solution) ApproxEq(b Solution) bool {
	if len(a) != len(b) {
		return false
	} else if len(a) == 0 {
		return true
	} else if len(a) == 1 {
		return ApproxEq(a[0], b[0])
	}
	return (ApproxEq(a[0], b[0]) && ApproxEq(a[1], b[1])) ||
		(ApproxEq(a[0], b[1]) && ApproxEq(a[1], b[0]))

}
